rclpy/src/rclpy/_rclpy.c

```diff
@@ -1333,6 +1333,96 @@ rclpy_create_publisher(PyObject * Py_UNUSED(self), PyObject * args)
   return PyCapsule_New(publisher, "rcl_publisher_t", NULL);
 }
 
+/// Destructor for an event.
+void
+_rclpy_destroy_event(PyObject * pycapsule)
+{
+  rcl_event_t * event = PyCapsule_GetPointer(pycapsule, "rcl_event_t");
+  rcl_ret_t ret_event = rcl_event_fini(event);
+  PyMem_Free(event);
+  if (ret_event != RCL_RET_OK) {
+    PyErr_Format(PyExc_RuntimeError,
+      "Failed to fini 'rcl_event_t': %s", rcl_get_error_string().str);
+    rcl_reset_error();
+  }
+}
+
+/// Create an event object for QoS event handling.
+/**
+  * This function will create an event handle for the given Subscription or Publisher parent.
+  *
+  * Raises MemoryError if the event can't be allocated
+  * Raises RuntimeError on initialization failure
+  * Raises TypeError if the capsules are not the correct types
+  *
+  * \param[in] pyevent_type Enum value of
+  *   rcl_publisher_event_type_t or rcl_subscription_event_type_t, chosen by the type of pyparent
+  * \param[in] pyparent Capsule containing the parent Publisher or Subscription
+  * \return capsule containing rcl_event_t
+  * \return NULL on failure
+  */
+static PyObject *
+rclpy_create_event(PyObject * Py_UNUSED(self), PyObject * args)
```

Comment:
Big function - can we split this?

Reply:
C is pretty verbose - do you really think this is actually more than one logical unit of code?

Reply:
Here is a re-implementation of your function (coding style, passing precisely the args etc. is left as an exercise for the reader :)

```c++
bool _is_pycapsule_publisher(PyObject* pyobject) {
  return PyCapsule_IsValid(pyparent, "rcl_subscription_t");
}

bool _is_pycapsule_subscription(PyObject* pyobject) {
  return PyCapsule_IsValid(pyparent, "rcl_publisher_t");
}

rcl_event_t * _allocate_and_zero_initialize_rcl_event() {
  rcl_event_t * event = (rcl_event_t *)PyMem_Malloc(sizeof(rcl_event_t));
  if (!event) {
    PyErr_Format(PyExc_MemoryError, "Failed to allocate memory for event");
    return NULL;
  }
  *event = rcl_get_zero_initialized_event();
  return event;
}

// forcing yourself to divide your code in smaller functions will help build
// generally useful utility functions such as this one instead of considering
// the code "write-only".
bool _check_rcl_errors_status() {
  if (ret != RCL_RET_OK) {
    rcl_reset_error();
    PyErr_Format(PyExc_RuntimeError, rcutils_get_error_string());
    return true;
  } else {
    return false;
  }
}

bool _initialize_event_publisher(event) {
  const rcl_ret_t ret = rcl_subscription_event_init(event, subscription, pyevent_type);
  return _check_rcl_error_status();
}

bool _initialize_event_subscription(event) {
  const rcl_ret_t ret = rcl_subscription_event_init(event, subscription, pyevent_type);
  return _check_rcl_error_status();
}

bool _initialize_event(event) {
  if(_is_pycapsule_publisher(event)) {
    return _initialize_event_publisher();
  } else if(_is_pycapsule_subscription(event)) {
    return _initialize_event_subscription();
  }
}

static PyObject *
rclpy_create_event(PyObject * Py_UNUSED(self), PyObject * args) {
  const rcl_event_t * event = _allocate_and_zero_initialize_rcl_event();
  if (!event) {
    return NULL;
  }
  if (_initialize_event(event)) {
    return event;
  } else {
    PyMem_Free(event);  // Nesting functions will help you ensuring cleanup code is needed only ONCE.
    return NULL;
  }
}
```